### ISSUE TITLE
[gemma-4]: attention accuracy issues for long context length (seq_len>=4096)

### ISSUE THREAD

**Issue by ssanjayTT** [tenstorrent/tt-metal#46211]
### Component / Area

model

### Issue Type (optional)

None

### Observed

On **Gemma4 text encoder** attention prefill unit tests (`models/demos/gemma4/tests/unit/test_attention.py::test_attention_prefill`), **global (`full_attention`) layers** drop below 0.99 PCC as sequence length grows. **Sliding (`sliding_attention`) layers remain ≥ 0.99** at all tested lengths, including 8192.

Reproduced on **QB2, Blackhole 1×4**, model **`google/gemma-4-31B-it`**, 2026-06-08:

| layer_type | seq_len | PCC   | threshold used | result @ 0.99 |
|------------|---------|-------|----------------|---------------|
| sliding    | 128     | 0.996 | 0.99           | ✅ |
| sliding    | 1024    | 0.992 | 0.99           | ✅ |
| sliding    | 4096    | 0.991 | 0.99           | ✅ |
| sliding    | 8192    | 0.991 | 0.99           | ✅ |
| global     | 128     | 0.994 | 0.99           | ✅ |
| global     | 1024    | 0.988 | 0.97*          | ❌ |
| global     | 4096    | 0.984 | 0.97*          | ❌ |
| global     | 8192    | 0.983 | 0.99           | ❌ (test FAIL) |

\*Lowered threshold from `models/demos/gemma4/tests/pcc_thresholds.json` — test passes CI despite PCC < 0.99.

Global PCC degrades monotonically with length (0.988 → 0.984 → 0.983). The **4096** case matches the original ~**0.984** report. Adding **8192** to `PREFILL_BUCKETS` exposes a new failure: `test_attention_prefill[blackhole-seq8192-global-1x4]` fails against the default 0.99 threshold (no lowered entry in `pcc_thresholds.json` yet).

### Expected

PCC should remain >=0.99

### 1. Steps (exact commands)

```bash
cd /path/to/tt-metal
source python_env/bin/activate
unset TT_VISIBLE_DEVICES TT_MESH_GRAPH_DESC_PATH
export HF_MODEL=google/gemma-4-31B-it
export TT_ARCH=blackhole

# PREFILL_BUCKETS = [128, 1024, 4096, 8192]
LOG=/tmp/gemma4_attn_prefill_buckets_31b_1x4.log
pytest models/demos/gemma4/tests/unit/test_attention.py::test_attention_prefill \
  -k "1x4" -v -s --timeout=7200 2>&1 | tee "$LOG"

# Or isolate long-context global failures:
pytest models/demos/gemma4/tests/unit/test_attention.py::test_attention_prefill \
  -k "1x4 and global and (seq4096 or seq8192)" -v -s --timeout=7200
```

### 2. Input data / link or description

- Model: `google/gemma-4-31B-it` via `HF_MODEL`
- Random Gaussian activations vs HuggingFace `Gemma4TextAttention` reference
- Global layer: `find_layer_idx(hf_text_config, "full_attention")` (layer 5 on 31B)
- Sliding layers: HF-compatible sliding-window + causal mask
- Global layers: full causal SDPA over entire `seq_len`
- TT path: bfloat16 QKV → per-head RMSNorm → RoPE → `ttnn.transformer.scaled_dot_product_attention` (`scale=1.0`) → o_proj

### 3. Frequency

consistently

### 1. Software Versions

- tt-metal: `main` @ `514951d985a7b0d5e5bff28accd86436b1318869`
- Python 3.10.19, pytest 9.0.3
- KMD 2.5.0, firmware bundle 19.6.0 (UMD notes 19.6.0 > latest fully tested 19.5.0 for BH)

### 2. Hardware Details

BH QB2 — Blackhole, 4-device mesh (1×4), chips `{0,1,2,3}`, IOMMU disabled

### Is this a regression?

Unknown

### Logs & Diagnostics

Full log: `/tmp/gemma4_attn_prefill_buckets_31b_1x4.log`

PCC summary (8 tests, ~72s):

```
PCC check: 0.9961367152600825 (threshold=0.99) [PASS]   # seq128 sliding
PCC check: 0.9940685144222294 (threshold=0.99) [PASS]   # seq128 global
PCC check: 0.9924499561026838 (threshold=0.99) [PASS]   # seq1024 sliding
PCC check: 0.9879827378816942 (threshold=0.97) [PASS]   # seq1024 global
PCC check: 0.9914382027412925 (threshold=0.99) [PASS]   # seq4096 sliding
PCC check: 0.9844741755687216 (threshold=0.97) [PASS]   # seq4096 global
PCC check: 0.9913127017643516 (threshold=0.99) [PASS]   # seq8192 sliding
PCC check: 0.9826218115631707 (threshold=0.99) [FAIL]  # seq8192 global
```

Pytest: **7 passed, 1 failed** — only `test_attention_prefill[blackhole-seq8192-global-1x4]`.

### Priority

None

### Impact

- **CI:** `seq4096-global` passes via lowered 0.97 threshold; `seq8192-global` now fails strict 0.99 (will fail CI unless threshold added or SDPA fixed)
- **Inference:** sliding layers (5 of every 6) appear fine; global layers (1 of every 6) accumulate error on long prefills (≥1024 tokens), potentially affecting logits on long prompts
- **Vision:** unverified until TT vision encoder unit tests exist

---
